sql: generalize orderingInfo to allow groups of columns

RaduBerinde opened the pull request:
This PR introduces the generalization of `orderingInfo` to allow column groups. This is necessary to be able to nest merge-joins, more details below and in the comments.

We don't yet use this new feature (or add tests covering it); I am posting this early to get some feedback as to whether doing this is reasonable for 1.1.

#### sql: use FastIntSet for orderingInfo.exactMatchCols

This is not motivated by a performance concern; we are about to support column
groups in `orderingInfo`, and every entry in `ordering` will be a `FastIntSet`.

#### util: FastIntSet optimizations, add Next, Equals

Adding Next and Equals primitives.

Rewriting the `ForEach` loop to make it more efficient. The main improvement is
that we exit right after the last set bit (`v` becomes 0).

Also improving the relevant test code.

#### sql: generalize orderingInfo to allow groups of columns

When dealing with ordering guarantees for joins (especially merge-joins), we
have a new problem: say we are doing a join for
```sql
SELECT t1.a, t1.b, t2.x, t2.y FROM t1 JOIN t2 ON t1.a = t2.x AND t1.b = t2.y
```
If the two tables have an ordering on `a,b` (respectively `x,y`), the result of
an inner join preserves this ordering (even a hash-join, because that preserves
the ordering of one of the streams). The problem is that we can't simply express
this ordering: any one of `t1.a,t1.b`, `t1.a,t2.y`, `t2.x,t1.b`, `t2.x,t2.y` are
equally valid (and some may be more useful than others). What we really want to
express is that the results are ordered first by `t1.a` AND AT THE SAME TIME
`t2.x`, and ties are broken by `t1.b` and at the same time `t2.y`.

This change generalizes orderingInfo to store orders of column groups. In this
commit, we don't yet use this feature and we don't yet extend the `orderingInfo`
tests to cover the new functionality; this will come in a follow-up commit.

The intention of the change was to retain current functionality unchanged;
however, two things were fixed in the process of rewriting code:
 - we now trim unnecessary orderings for ordinality nodes;
 - a `nosort` node preserves the `unique` flag of the underlying ordering when
   appropriate.

#### sql: rename exact-match columns to constant columns

We rename the concept of "exact-match columns" to "constant columns" which
is a much better name.

#### sql: ordering fixes and tests
    
Improve the match and trim functions to detect redundant columns. This wasn't an
interesting case before: it only applied in odd cases when we had desired
orderings with repeated columns (e.g. 1+,2-,1+). But with column groups an
ordering like 1+,2+,3+ has a redundant column if 1 and 3 are in the same group.

Add tests with column groups and a new test that verifies the trim guarantee on
random inputs.

#### sql: rename orderingInfo.unique to isKey
    
Renaming the concept of "unique" to the slightly more general concept of "key",
as defined in https://pdfs.semanticscholar.org/7834/6392b3d157173d7a3b15869e3c107ca3b715.pdf

#### sql: use orderingInfo column groups for ordinalityNode
    
When possible, we put the ordinality node in the first column group in the
ordering.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17439)
<!-- Reviewable:end -->

petermattis commented on pkg/util/fast_int_set.go:
```diff
@@ -64,25 +75,108 @@ func (s *FastIntSet) Empty() bool {
+// Note: this is efficient for small sets, but each call takes linear time for large sets.
+func (s *FastIntSet) Next(startVal uint32) (uint32, bool) {
+	if startVal < smallValCutoff {
+		for v := s.smallVals >> startVal; v > 0; {
```
FYI, https://tip.golang.org/pkg/math/bits/ is coming in go1.9. These are supported as intrinsics in the compiler so the portable versions are never used (AFAIK).

petermattis replied:
Also, https://godoc.org/golang.org/x/tools/container/intsets#Sparse might be of interest.

RaduBerinde replied:
On `bits`, yeah, I have a TODO somewhere to switch to those.

Thanks for the pointer on `Sparse`, that is very useful! I will look at switching to that.

RaduBerinde replied:
Unfortunately it looks like Sparse always allocates a block unless it's not empty. Which is pretty lame, it could be using the root block.

RaduBerinde replied:
Filed https://github.com/golang/go/issues/21311

RaduBerinde commented:
Alberto pointed out this paper: https://pdfs.semanticscholar.org/7834/6392b3d157173d7a3b15869e3c107ca3b715.pdf

I think that structure is what we want to implement longer-term; if we decide we don't want this change in 1.1, I will give up this change and instead work towards implementing that for 1.2.

arjunravinarayan commented:
I like this change quite a bit (reviewed r1 and r3 closely, skimmed r2). The comments are helpful, but it took me a while to understand the pattern (a discussion offline with @raduberinde took place). I would love to get this in 1.1.

I worked off of this commit to experiment with whether column groups can help with propagating orderings after JOINs, and it seems to work very well (WIP PR submitted). If you agree that that commit can be cleaned up, I'd strongly argue to get both the PRs in 1.1, as this would help greatly with our performance for more complex queries with JOINs.

I'm very interested in reading @knz's review as well.

---

Reviewed 6 of 6 files at r1, 1 of 4 files at r2, 8 of 11 files at r3, 2 of 2 files at r4.
Review status: 9 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsql_physical_planner.go, line 606 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-Kqi-Mz4tkheELIuNy4Y:-Kqi-Mz4tkheELIuNy4Z:b-u3oad4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6debbdd0e435d94e533559525f10b685b0cf14/pkg/sql/distsql_physical_planner.go#L606)):*
> ```Go
> 		Columns: make([]distsqlrun.Ordering_Column, 0, len(planOrdering)),
> 	}
> 	for _, group := range planOrdering {
> ```

Add comment: trying to find any column in the group that is part of the processor output.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-:-KqhzoUf7ALEVZWMrmNX:boeei37)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 9 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 606 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-Kqi-Mz4tkheELIuNy4Y:-KqiwokA1WhKN-fnZ38c:b-dtkds3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6debbdd0e435d94e533559525f10b685b0cf14/pkg/sql/distsql_physical_planner.go#L606)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Add comment: trying to find any column in the group that is part of the processor output.
</blockquote></details>

Done, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439)*
<!-- Sent from Reviewable.io -->

knz commented on pkg/util/fast_int_set.go:
```diff
@@ -17,10 +17,11 @@ package util
+// allocations when the values are small. It is not thread-safe.
 type FastIntSet struct {
 	smallVals uint64
 	largeVals map[uint32]struct{}
```
So given the range of values that's likely to occur, from experience this is much more efficient:
- `largeVals []int64`
- set with `largeVals[i/64] |= 1 << (i%64)`
- check with `largeVals[i/64] & ( 1 << (i%64) )`
- you can also skip 8 bits at a time in the same way you do for smallVals

RaduBerinde replied:
This would only matter when we go beyond 64 columns.. I think we will probably switch to `intsets.Sparse` here anyway (I'm working on some improvements on that code).

knz replied:
wel ok but still I'm not keen on keeping a map here until it goes away.

RaduBerinde replied:
Why? It's not allocated unless we actually need it. And there are usecases of FastIntSet where your solution is terrible (inserting a few large values, which we conceivably could do in other places where we use this structure, e.g. we use it for `ColumnIDs` in another context and there are no guarantees that IDs will always be small).

knz replied:
All right, but you could still have a map[int32]int32  and pack multiple bits per map entry instead of map[int64]struct{}.

RaduBerinde replied:
Yes, but it is not trivial to implement correctly and it would be throwaway work because we probably do want to switch to the `Sparse` implementation, and it would utterly useless for this change because we normally don't have 64 columns, and if we do, I'm willing to bet the "hot path" of such a query won't be representing column groups in this structure :)

knz replied:
Ack, let's let this sleep then.

knz commented on pkg/sql/logictest/testdata/logic_test/needed_columns:
```diff
@@ -67,7 +67,7 @@ query ITTTTT
 ----
 0  render      ·  ·  ("1")                       ="1"
-1  ordinality  ·  ·  (s[omitted], "ordinality")  +"ordinality",unique
+1  ordinality  ·  ·  (s[omitted], "ordinality")  ·
```
Now that you can represent multiple orderings, why not use that for ordinalityNode too then?

RaduBerinde replied:
Yeah I plan to do this in a follow-up commit.

knz commented on pkg/sql/ordering.go:
```diff
@@ -24,35 +24,83 @@ import (
-//    b, d, a
-//    a, b, c, d
-//    b, a, c, d
+// == Exact match columns ==
```
I think this PR may also be the opportunity to rename this concept from "exact match" (which is confusing) to "known-constant" (which actually reflects the property).

For example `SELECT 42 AS x, k FROM kv` will have x known-constant, although there is no "match" to speak of.

RaduBerinde replied:
Good point. I added a commit - I called them "constant columns", should be more clear now.

knz replied:
Thanks!

knz commented on pkg/sql/ordering.go:
```diff
@@ -24,35 +24,83 @@ import (
+// and we know that each row has a unique set of values on these columns. For
+// example, if we have a unique index on columns (a, b), the results have the
+// ordering a+,b+ and the unique flag tells us that this satisfies any desired
+// ordering that has a+,b+ as a prefix (such as a+,b+,c-,d+).
```
Thanks for explaining this. Is it perhaps worth investigating having renderNode check the expressions to know which expressions produce unique results? For standalone constants or indexedvars it's trivial, but I think we could also annotate built-in functions to indicate they are [injective](https://en.wikipedia.org/wiki/Injective_function) and thus preserve unicity of their arguments, and propagate this during type checking. What do you think?

RaduBerinde replied:
Yes I think we should have that kind of infrastructure. It will be even more useful when we implement a better structure which allows multiple "keys" (like the one described in the paper above).

knz replied:
Filed as https://github.com/cockroachdb/cockroach/issues/17467

knz reviewed (approved):
I'm happy to see this idea we discussed two months ago to be finally implemented. This is great. Thanks!

RaduBerinde commented:
I have added more commits and updated the description. This PR is ready for final review.

RaduBerinde committed:
sql: rename orderingInfo.unique to isKey

Renaming the concept of "unique" to the slightly more general concept of "key",
as defined in https://pdfs.semanticscholar.org/7834/6392b3d157173d7a3b15869e3c107ca3b715.pdf

RaduBerinde committed:
sql: ordering fixes and tests

Improve the match and trim functions to detect redundant columns. This wasn't an
interesting case before: it only applied in odd cases when we had desired
orderings with repeated columns (e.g. 1+,2-,1+). But with column groups an
ordering like 1+,2+,3+ has a redundant column if 1 and 3 are in the same group.

Add tests with column groups and a new test that verifies the trim guarantee on
random inputs.

RaduBerinde committed:
sql: rename exact-match columns to constant columns

We rename the concept of "exact-match columns" to "constant columns" which
is a much better name.

RaduBerinde committed:
sql: generalize orderingInfo to allow groups of columns

When dealing with ordering guarantees for joins (especially merge-joins), we
have a new problem: say we are doing a join for
```sql
SELECT t1.a, t1.b, t2.x, t2.y FROM t1 JOIN t2 ON t1.a = t2.x AND t1.b = t2.y
```
If the two tables have an ordering on `a,b` (respectively `x,y`), the result of
an inner join preserves this ordering (even a hash-join, because that preserves
the ordering of one of the streams). The problem is that we can't simply express
this ordering: any one of `t1.a,t1.b`, `t1.a,t2.y`, `t2.x,t1.b`, `t2.x,t2.y` are
equally valid (and some may be more useful than others). What we really want to
express is that the results are ordered first by `t1.a` AND AT THE SAME TIME
`t2.x`, and ties are broken by `t1.b` and at the same time `t2.y`.

This change generalizes orderingInfo to store orders of column groups. In this
commit, we don't yet use this feature and we don't yet extend the `orderingInfo`
tests to cover the new functionality; this will come in a follow-up commit.

The intention of the change was to retain current functionality unchanged;
however, two things were fixed in the process of rewriting code:
 - we now trim unnecessary orderings for ordinality nodes;
 - a `nosort` node preserves the `unique` flag of the underlying ordering when
   appropriate.

RaduBerinde committed:
util: FastIntSet optimizations, add Next, Equals

Adding Next and Equals primitives.

Rewriting the `ForEach` loop to make it more efficient. The main improvement is
that we exit right after the last set bit (`v` becomes 0).

Also improving the relevant test code.

RaduBerinde committed:
sql: use FastIntSet for orderingInfo.exactMatchCols

This is not motivated by a performance concern; we are about to support column
groups in `orderingInfo`, and every entry in `ordering` will be a `FastIntSet`.

knz commented:
Mostly good to go, just a minor comment nit and a suggestion for ordinalityNode.

---

Reviewed 6 of 6 files at r1, 1 of 4 files at r2, 7 of 15 files at r6, 4 of 4 files at r7, 11 of 11 files at r8, 7 of 7 files at r9, 2 of 2 files at r10, 12 of 12 files at r11, 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 64 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-Kr1OVgoOCSxpvEBamL5:-Kr1OVgpoXVgFh-BwqF3:bg9ecuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ae289ca8c3273abd2750429184fe3b23fe14f47/pkg/sql/ordering.go#L64)):*
> ```Go
> // A set of columns S forms a "key" if any two rows that are equal on S are
> // necessarily equal on all columns. A case of this is when we have a unique
> // index on the columns in S.
> ```

Extend with: "Conversely, no two rows compare equal on all the columns in the set, i.e.
they are unique when projecting them along these columns."

---

*[pkg/sql/ordinality.go, line 128 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-Kr1P7an6I-ZyFr792aW:-Kr1P7aoMH_s09vTz4dt:bq1ui6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/02371fb4e59787da54c9748af8a38448950411f4/pkg/sql/ordinality.go#L128)):*
> ```Go
> 	// change. So here retrieve the ordering of the source again.
> 	o.ordering = planOrdering(o.source).copy()
> 	if len(o.ordering.ordering) > 0 {
> ```

Here I think the ordering column can be appended to *every* group that's ascending, not just the first one.

Then, if the ordering column cannot be appended to any group, then a new group can be crated just for it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-:-Kr1PSDJ0FzMIJt0q0Ma:bawqmap)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

---

Review status: 0 of 21 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/ordering.go, line 64 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-Kr1OVgoOCSxpvEBamL5:-Kr3jrqXOSHgCrOAsDjC:b-ewj6ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ae289ca8c3273abd2750429184fe3b23fe14f47/pkg/sql/ordering.go#L64)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Extend with: "Conversely, no two rows compare equal on all the columns in the set, i.e.
they are unique when projecting them along these columns."
</blockquote></details>

This doesn't need to be the case - this is why the key concept is more general. We can have multiple rows that are equal along those columns, as long as those rows are equal on all columns (same consequences on relative ordering). This might mean we are able to propagate the "key" property in more cases than the "unique" property (e.g. a join in some cases).

---

*[pkg/sql/ordinality.go, line 128 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-Kr1P7an6I-ZyFr792aW:-Kr3kdx5PRPT1-a6My4I:bqp5co8) ([raw file](https://github.com/cockroachdb/cockroach/blob/02371fb4e59787da54c9748af8a38448950411f4/pkg/sql/ordinality.go#L128)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Here I think the ordering column can be appended to *every* group that's ascending, not just the first one.

Then, if the ordering column cannot be appended to any group, then a new group can be crated just for it.
</blockquote></details>

This is smart, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-:-Kr3n9fth84rufMsr9y6:b-814g8k)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r14, 3 of 11 files at r15, 1 of 7 files at r16, 11 of 12 files at r18, 2 of 2 files at r19.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 64 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-Kr1OVgoOCSxpvEBamL5:-Kr6O9dp8PT_EZUF_MNB:bmf28qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ae289ca8c3273abd2750429184fe3b23fe14f47/pkg/sql/ordering.go#L64)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This doesn't need to be the case - this is why the key concept is more general. We can have multiple rows that are equal along those columns, as long as those rows are equal on all columns (same consequences on relative ordering). This might mean we are able to propagate the "key" property in more cases than the "unique" property (e.g. a join in some cases).
</blockquote></details>

Oh I see that's insightful! Mind explaining this in the comment then :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-:-Kr6ObDrAo1elRmQxd_q:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 18 of 21 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/ordering.go, line 64 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-Kr1OVgoOCSxpvEBamL5:-Kr6t8oPJ4TSyp03mTLP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ae289ca8c3273abd2750429184fe3b23fe14f47/pkg/sql/ordering.go#L64)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Oh I see that's insightful! Mind explaining this in the comment then :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: accept more matches in computeMergeJoinOrdering

Fixing an omission in computeMergeJoinOrdering; it now generates merge joins in
more cases.

RaduBerinde commented:
I pushed another commit that makes `computeMergeJoinOrdering` return better matches in some cases. Unfortunately, I think that some of these cases only work if column groups are "equivalency groups". For example, an ordering on `(1/3)+, (2/4)+` satisfies a desired ordering of `1+,2+,3-,4-` if `1/3` are an equivalency group.

Currently, the only case where we have groups that are not equivalency groups is ordinalityNode. In fact, I believe what we did there doesn't quite fit the current definition either, which says that two columns in the same group can be used interchangeably in any ordering. If we have an input ordered by `1+,2+` and the ordinality column is 3, we expose ordering `(1,3)+,2+`. But 1 and 3 can't be used interchangeably: an ordering of `1+` does not necessarily also guarantee an ordering on `3+` (there can be multiple rows with the same value on `1`, and among such a group there is no guarantee on `3`).

So I would like to redocument column groups as equivalency groups (which is what they do in the paper above too) and revert the changes we made for ordinalityNode. Does this sound good @knz?

RaduBerinde commented:
Updated. Removed commit with changes to ordinalityNode and documented columns in groups as being equal.

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r24, 4 of 11 files at r25, 10 of 12 files at r28, 2 of 2 files at r29.
Review status: 20 of 21 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17439#-:-Kr7r0HEwcRuxtcz5yDp:bnfp4nl)*
<!-- Sent from Reviewable.io -->
